docs/proposals/2019-02-08 - Fallback Content Elements.md

```diff
@@ -0,0 +1,90 @@
+# Alternatives to Raw HTML
+
+# Problem
+
+The content element `raw_html` is often used to inject Javascript into an ANS document which makes it difficult for downstream applications to render natively. Even if the application tries to render it as a embeded browser, there is no guarentee that it will render correctly in all applications. And some applications (ex Apple TV or $50 Kindle) may not have the input or processing capabilities needed to effectively render the experience.
+
+# Proposal
+
+Add a required `alternative` trait to the `raw_html` content element, that contains an array of one content element (allows for multiple later) that can notionally replace the `raw_html` block. The content in the array can only be of type `text`, `image`, `gallery` or `video`. When `upvert`ing this trait should be filled this with a `text` element that has a stock `This content is not available on this platform` message (or similar).
```

Comment:
This trait makes sense, where it can be used.  But it should NOT be required.

Reply:
Requiring this forces, or at least encourages, content creators to add affordable for non-html clients.

Reply:
I agree with Chad that this field being required is important. Without it there is no way to make sure that the issues with `raw_html` are mitigated. If content producers are unhappy with the default content here then we can change default, but in practice they should be using Ellipsis to tailor it for each case, whether that is a message, link, video, image, or whatever. Having the content not show up at all is worse than seeing a stock message in almost every case. If it's not required, then everyone is going to forget to set this field and that defeats the purpose. I don't think it is unreasonable to tell users that if they use `raw_html` it comes at the cost of making sure there is a viable alternative.

Reply:
@robpc @chadmoone Are we comfortable saying, then, that any legacy content that has already been saved contains raw_html (of which there is quite a bit) cannot be changed or re-published until an alternative is manually added to each raw_html object? (I am, but it's a tradeoff worth being explicit about.)

What about ingested content from legacy CMSes that store data in HTML (which is almost all of them?) When an automated ingestor pulls this old content in, and can't parse the entire article into ANS elements, what is the correct behavior? 

I suspect the result you'd be incentivizing is that any content where this occurs ends up with a very generic "Missing Element" placeholder image. It's not immediately clear to me that this is better than simply skipping over the raw_html element.

Reply:
The reason I think it's better is that it's explicit. I think it applies to both newly generated content and imported content.  Right now, what happens in apps is that there is just nothing there at all.

To the end user, I think a clear indicator that there is article content here which can't be rendered on this device is important, because they might be able to try reading it another way. But just as important is internal troubleshooting. I think many of the Arc partners have a more hands-off approach to apps, but this is absolutely a common case on WP side. We get asked all the time which a given chunk of content doesn't show up on the app, or even if we don't hear about it, we notice it ourselves when the article body references something which is nowhere to be found.

Admittedly, this is a crap experience when the bulk of the content elements are actually just `raw_html`.  In this case however, I think I'd recommend:
1. Force at least the default `alternative` value, but just as importantly
2. Set the new `rendering_guides.preferred_method` defined in #198 appropriately, which would cause supporting clients which cannot render `raw_html` to open in a web view (if available).

I hadn't considered this case before, but for the bulk ingestion of purely HTML content which is just going to get dumped primarily/completely into `raw_html` elements (😢), then I think both of the above should be applied.  For legacy content already imported which this applies to, I'm not sure if there is a good way to update the second part retroactively, but maybe someone else has an idea there.

Reply:
If part of the recommendation for legacy (or otherwise substantially HTMLish) content is to set `rendering_guides.preferred_method`:`www`, what value does setting the `alternative` add? It seems like once `www` is set, there's no real choice but to open a webview anyway.
  
My worry here is that if by making this field required, it may actually have the opposite effect as intended. The intended effect, AFAICT, is that for each piece of content where someone has decided to manually insert some HTML extract, they should be asked to set a replacement content simultaneously. In this way, non-HTML renderers can insert something in place of the HTML having been safely assured that they are showing what was intended by producer. So far, so good. 

But imagine the human equivalent of the ingestor scenario. A user is asked to (intentionally) insert some HTML in their article because of $BUSINESS_REASONS. Then we inform that they can't publish the article without a replacement, even if there is no reasonable replacement available. So the user does what? My guess is they select "text" mode and input "N/A" or the empty string or whatever it takes to make the error message go away. At that point, the field no longer captures "intended replacement" in a meaningful way, and the apps are back to rendering content with potentially critical stuff missing.

Is it possible that this is less a data model requirement and more a business-logic requirement: that for all *new* content created, either `rendering_guides.preferred_method`:`www` must be set, OR an alternative should be set for each-and-every `raw_html`? This gives legacy ingestors a clean path forward (just set `www` on everything) while also pushing users towards cleaner content.

Reply:
The "Classic App" (at the time, just "the iPhone app") used to run off of an RSS feed that didn't support any content type except text, images, and videos. In place of any other embed, it would render "This content is not supported." One of the first requests I worked on was getting rid of those error messages, because it made the app look broken.

I'm much more in favor of transparency than sweeping it under the rug, but I agree that mandating this doesn't scale well for outside and existing content.

What do you guys think about, if no alternatives are provided, replacing the raw HTML with a "view on web" embed? Then the default experience isn't a dead end, and engaged newsrooms can provide a better experience.

Reply:
> One of the first Newsroom requests I worked on was getting rid of those error messages, because it made the app look broken.

Silently hiding it has the same effect, or worse they don't notice. I don't think `This content is not supported` is the desired outcome but it makes what is happening clear and hopefully would bug them to use the content properly. The difference with your example is there was no path for the newsroom to correct it. Here if the newsroom is saying that `This content is not supported` is broken then the response is that they didn't properly set the alternative and they should go do that. 

I think these points are valid for not being draconian retroactively, but as @gengel says I am not sure a better way to do it. Without a forced alternative it is easy to fall into the same pattern we have now, where these items fall through the cracks and people are "shocked" to find out they are missing from some platforms. Giving a default message doesn't make it more broken, it just exposes to newsroom users that it is broken which is an improvement. Even better they have the tools to fix it. 

From the statements here it seems like people are imagining a situation where they add the `raw_html` and then an hour after publishing realize that there is a stock message saying the content doesn't work. That's not how this should work. Ellipsis or whatever other tool should be showing this to them as the alternative as they are adding the content. The default message is really just an invitation to put something better. 

Another thought I had over the weekend was having a flag that says the content is not required, which would allow the alternatives to not be provided and signal that it was safe to ignore the content. The upverting could default to this being turned on, but going forward it would be turned off and the default message would be provided. 

I don't love this because of the issues already mentioned, but those could be mitigated by clear messaging in the content editor. This solution would allow for the cases mentioned (sequential `raw_html` content, dynamically imported etc). It also allows the editor to indicate it's not needed, but requires them to make take an action to say it's not needed. There are many ways this could fall through so I am not really advocating for it, but it's the only alternative I can think of.

Reply:
> If part of the recommendation for legacy (or otherwise substantially HTMLish) content is to set rendering_guides.preferred_method:www, what value does setting the alternative add? It seems like once www is set, there's no real choice but to open a webview anyway.

@gengel I believe that @chadmoone is saying that you would set a default alternative for most situations, but in situations where that doesn't work (import of almost all html only content, etc) you could set the preferred view as another way to deal with it. These two requests are in our minds different ways of handling the same issue. One is a hammer, the other a scalpel.

Reply:
Whoops missed this statement from @gengel

> Is it possible that this is less a data model requirement and more a business-logic requirement: that for all new content created, either rendering_guides.preferred_method:www must be set, OR an alternative should be set for each-and-every raw_html? This gives legacy ingestors a clean path forward (just set www on everything) while also pushing users towards cleaner content.

I think that is what @chadmoone is saying and works for me

Reply:
@robpc Cool. In that case, I suggest we drop the `required` attribute for the JSON schema and instead forward that requirement Ellipsis (cc @awong7)

Reply:
I am giving approval based on the decision here to drop the required attribute from the schema. If this decision is changed I'd like to withdraw approval and discuss more.

Reply:
Updated the language to reflect @gengel and @chadmoone 's feedback. Should meet @styonsk's concerns, though I would like to point out that this element is still required even if within the spec we will mark it as not to support one very specific exception. 

```
This trait is required unless the `rendering_guides.preferred_method` is set to `website` only.
```

I have concerns about that, but as long as Ellipsis is willing to support the intent there then I am happy to move it forward.